### PR TITLE
Learner devices status update

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Turns the learner device status user field to a ForeignKey rather than OneToOne field to prevent errors when trying to create learner device statuses for the same learner for syncs from multiple devices
* Adds sync hook to selectively delete LearnerDeviceStatus to prevent trying to sync more than one LearnerDeviceStatus for a single user to another device.

## References
Fixes [#12043](https://github.com/learningequality/kolibri/issues/12043)

## Reviewer guidance
The logic is relatively straight forward, I think - I am just not sure how to do any sort of test of this change, especially the version sync hook.

QA specific guidance:

Install the PR asset onto a server device and onto two single user devices.

Setup syncing with those two single user devices to the server for the same learner account.

Setup 0.16.1 onto another server device, and import the facility.

Setup another single user syncing device using 0.16.1 and import the same learner again.

Ensure syncing continues unabated.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 2d91ec9f5a35fc65a39d5abf078afaa1f5082b8d  | 
|--------|

### Summary:
This PR updates the `LearnerDeviceStatus` model to support multiple device statuses per user and adds a sync operation to manage data consistency across device syncs.

**Key points**:
- Changed `LearnerDeviceStatus.user` from `OneToOneField` to `ForeignKey`.
- Added `LearnerDeviceStatusOperation` to handle deletion of problematic `LearnerDeviceStatus` records during sync.
- Introduced `LearnerDeviceStatusHook` to trigger the new operation during sync.
- Added migration `0022_learner_device_status_multiple_devices.py` for database schema update.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by rtibbles** [comment]
Hey @ellipsis, give me a code review

File: kolibri/core/device/kolibri_plugin.py [hunk 1]
```diff
@@ -64,3 +67,40 @@ def post_transfer(
...
+        # get the user_id for the single user sync
+        # if it's not a single user sync, this will be None
+        user_id = get_user_id_for_single_user_sync(context)
+
+        # get the instance_id of the remote instance
+        instance_id = (
+            context.sync_session.client_instance_id
+            if context.is_server
+            else context.sync_session.server_instance_id
+        )
```
Comment (bjester) [inline kolibri/core/device/kolibri_plugin.py]: So this `downgrade` method is called when this device is the producer of sync data. When it's a pull, then it must be the server if it's the producer, so restricting the learner status to only the `client_instance_id` device makes sense.

Although, when it's a push, in order for the device to only send one status, it can first consider whether the current device has created a sync status. If the current device has, we would still want to exclude the `client_instance_id`, because being a push producer (aka `not context.is_server`), the client's status is still what matters. If the current device hasn't created one, then we would want to delete all but the most recent user status.

So in summary, I don't think this code will function as expected when the device is pushing. For pulling, this looks okay.

When it's pushing without a user sub-partition, we would want to ensure the statuses are unique for each user, preferably using the most recent status for each user, meaning no explicit filtering on the instance. So the same approach for pulling doesn't quite work for pushing.

**Review by bjester — CHANGES_REQUESTED** [review]
More specific handling is required when the device is the pusher.

File: kolibri/core/device/kolibri_plugin.py (hunk 1 above)
Comment (rtibbles) [reply kolibri/core/device/kolibri_plugin.py]: Thanks @bjester - I didn't think I'd get it right first time. If you have any ideas on how to write a robust test case for this, so I can be sure of covering the bases, I'm very open to inspiration!

Comment (bjester) [reply kolibri/core/device/kolibri_plugin.py]: Honestly coming back to the conversation about this, it was difficult to wrap my head around it! I realized the logic isn't straightforward at all :melting_face: 

I was wondering about testing. Since this relies on different versions, it doesn't seem straightforward either. The `KolibriVersionedSyncOperationTestCase` demonstrates how you could possibly set up a unit test. For sync integration tests, I believe it could be possible to set one up overriding `CUSTOM_INSTANCE_INFO` which influences the version that morango uses and communicates during syncing. https://github.com/learningequality/kolibri/blob/a2f580d6d065602f33ae1dccc1663d66e8f9e99c/kolibri/core/auth/test/test_morango_integration.py#L293

Comment (rtibbles) [reply kolibri/core/device/kolibri_plugin.py]: Ah yeah, I had forgotten about the custom instance info, that does seem like the right way to go for the integration tests. Will take a look at the version sync tests too, thanks!

**Review by bjester — APPROVED** [review]
The latest updates make sense! This logic will make you loopy :loop:

**Comment by pcenov** [comment]
Hi @rtibbles no issues observed while checking the above described QA scenario - the devices were syncing correctly. I do see some errors in the server logs so posting these here for you in case you find anything that needs to be fixed:

[logs-and-db.zip](https://github.com/learningequality/kolibri/files/15323192/logs-and-db.zip)

**Comment by bjester** [comment]
> Hi @rtibbles no issues observed while checking the above described QA scenario - the devices were syncing correctly. I do see some errors in the server logs so posting these here for you in case you find anything that needs to be fixed:
> 
> [logs-and-db.zip](https://github.com/learningequality/kolibri/files/15323192/logs-and-db.zip)

Looks like some DB locking errors
